### PR TITLE
Web console: fix json input

### PR DESCRIPTION
**Pull request by vogievetsky**
Fixed and improved the JSON Input component (as seen in the Edit Spec step of the data loader)
Specifically fixed the issue with the JSON input sometimes moving the cursor when editing, fixed the undo stack.
The fix is based around maintaining the state of the JSON Input in a single state variable (`InternalValue`) and thus not letting it get desynchronized.

While at it I also added a nice error display in case there is an issue with the JSON

image

Clicking on it takes your cursor to the place with the error which is sweet - doing this required me to nest the Ace editor inside a div which causes a lot of snapshot diffs.

Also updated the snapshots for all the users of JSON Input to shallow render so in the future they would be immune to JSON Input structure changes.

This PR has:
- [x] been self-reviewed.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links
- [x] added unit tests or modified existing tests to cover new code paths
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: web-console/src/components/json-input/json-input.tsx [hunk 1]
```diff
@@ -48,52 +60,93 @@ interface JsonInputProps {
...
+        }}
+      />
+      {showErrorIfNeeded && internalValueError && (
+        <div
+          className="json-error"
+          onClick={() => {
+            if (!aceEditor.current || !internalValueError) return;
+
+            // Message would be something like:
+            // `Found '}' where a key name was expected at line 26,7`
```
[the remark this answers is not in the document]
Comment (gianm) [reply web-console/src/components/json-input/json-input.tsx:138]: I suppose this won't always work; is that ok?

File: web-console/src/components/json-input/json-input.tsx [hunk 2]
```diff
@@ -37,6 +38,17 @@ function stringifyJson(item: any): string {
   }
 }
 
+// Not the best way to check for deep equality but good enough for what we need
```
[the remark this answers is not in the document]
Comment (vogievetsky) [reply web-console/src/components/json-input/json-input.tsx:41]: Here is an in depth look at it https://www.mattzeunert.com/2016/01/28/javascript-deep-equal.html

TL;DR the method implemented here would not work for this case for example:

`JSON.stringify({a: 1, b: 2}) !== JSON.stringify({b: 2, a: 1})`

But that is ok because in practice the object output from the JSON input will be the same it later gets back. And if that check fails the worst that will happen is that the cursor would jump. It is also hard to find a decent object deep-equals library that does not come with a bunch of dependencies (as it tries to do a 100% thorough job) and a lot of file size.

File: web-console/src/components/auto-form/__snapshots__/auto-form.spec.tsx.snap [hunk 3]
```diff
@@ -2,521 +2,135 @@
 
 exports[`auto-form snapshot matches snapshot 1`] = `
 <div
-  class="auto-form"
```
[the remark this answers is not in the document]
Comment (vogievetsky) [reply web-console/src/components/auto-form/__snapshots__/auto-form.spec.tsx.snap:5]: That is correct.

Here is how I review them:

```
The snapshot change is either tiny or structural (huge) there is no in between.

If the change is tiny:
  Review the changed lines and understand what caused that change. Was a class or some other property added / removed? Does it make sense?

If the change is structural:
  Do not review the individual lines, only review which snapshot files changed. Does it make sense that the snapshot of this component changed so drastically?
```

In this case all the changes are structural. I have wrapped the JSON Input in an extra `div` (for good reason) which causes all the snapshots of its inner workings to be indented by an extra level. This causes huge diffs. Seeing as I will have huge changed in these snapshots anyway I changed them to be shallow rendered (which they should have been from the start - but better late than never). This explains the structural nature of the change. You are free to comment on whether or not you think it is a good idea. (see below)

File: web-console/src/components/auto-form/auto-form.spec.tsx [hunk 4]
```diff
@@ -16,14 +16,14 @@
  * limitations under the License.
  */
 
-import { render } from '@testing-library/react';
+import { shallow } from 'enzyme';
```
[the remark this answers is not in the document]
Comment (vogievetsky) [reply web-console/src/components/auto-form/auto-form.spec.tsx:19]: Yup, this changes the virtual rendered from a DOM renderer `import { render } from '@testing-library/react';` to a shallow react component level renderer `import { shallow } from 'enzyme';` for the snapshots.

This change makes sense because the auto-form does not really make any of its own cool DOM it only assembles other components so testing their deep structure should not be its concern - otherwise it will have structural changes in the snapshots every-time someone changes the JSON Input

File: web-console/src/components/json-input/json-input.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (vogievetsky) [reply web-console/src/components/json-input/json-input.tsx:138]: Yeah, it it does not find `/line (\d+),(\d+)/` it would just not do anything due to the `if (!m) return;` which is cool I think. But as I write this I realize it would be great to add a test to make sure that if the error format changes in the Hjson library in a future version we catch that - will do.

[the remark this answers is not in the document]
Comment (vogievetsky) [reply web-console/src/components/json-input/json-input.tsx]: Done https://github.com/apache/druid/pull/10271/commits/8eb77e92e23487a0c391e770bcec010545758c6e
